MantidQt/API/src/NonOrthogonal.cpp

```diff
@@ -0,0 +1,227 @@
+}
+
+void stripMatrix(Mantid::Kernel::DblMatrix &matrix) {
+  std::size_t dim = matrix.Ssize() - 1;
+  Mantid::Kernel::DblMatrix temp(dim, dim);
+  for (std::size_t i = 0; i < dim; i++) {
+    for (std::size_t j = 0; j < dim; j++) {
+      temp[i][j] = matrix[i][j];
+    }
+  }
+  matrix = temp;
+}
+
+template <typename T>
+void doProvideSkewMatrix(Mantid::Kernel::DblMatrix &skewMatrix, T workspace) {
+  // The input workspace needs have
+  // 1. an HKL frame
+  // 2. an oriented lattice
+  // else we cannot create the skew matrix
+  const auto &sample = workspace->getExperimentInfo(0)->sample();
+  const auto &run = workspace->getExperimentInfo(0)->run();
+  auto specialCoordnateSystem = workspace->getSpecialCoordinateSystem();
+  checkForSampleAndRunEntries(sample, run, specialCoordnateSystem);
+
+  // Create Affine Matrix
+  Mantid::Kernel::Matrix<Mantid::coord_t> affineMatrix;
+  try {
+    auto const *transform = workspace->getTransformToOriginal();
+    affineMatrix = transform->makeAffineMatrix();
+  } catch (std::runtime_error &) {
+    // Create identity matrix of dimension+1
+    std::size_t nDims = workspace->getNumDims() + 1;
+    Mantid::Kernel::Matrix<Mantid::coord_t> temp(nDims, nDims, true);
+    affineMatrix = temp;
+  }
+
+  // Extract W Matrix
+  auto wMatrixAsArray =
+      run.template getPropertyValueAsType<std::vector<double>>("W_MATRIX");
+  Mantid::Kernel::DblMatrix wMatrix(wMatrixAsArray);
+
+  // Get the B Matrix from the oriented lattice
+  const auto &orientedLattice = sample.getOrientedLattice();
+  Mantid::Kernel::DblMatrix bMatrix = orientedLattice.getB();
+  bMatrix *= wMatrix;
+
+  // Get G* Matrix
+  Mantid::Kernel::DblMatrix gStarMatrix = bMatrix.Tprime() * bMatrix;
+
+  // Get recalculated BMatrix from Unit Cell
+  Mantid::Geometry::UnitCell unitCell(orientedLattice);
+  unitCell.recalculateFromGstar(gStarMatrix);
+  skewMatrix = unitCell.getB();
+
+  // Provide column normalisation of the skewMatrix
+  normalizeColumns(skewMatrix);
+
+  // Setup basis normalisation array
+  std::vector<double> basisNormalization = {orientedLattice.astar(),
+                                            orientedLattice.bstar(),
+                                            orientedLattice.cstar()};
+
+  // Expand matrix to 4 dimensions if necessary
+  if (4 == workspace->getNumDims()) {
+    basisNormalization.push_back(1.0);
+    Mantid::Kernel::DblMatrix temp(4, 4, true);
+    for (std::size_t i = 0; i < 3; i++) {
+      for (std::size_t j = 0; j < 3; j++) {
+        temp[i][j] = skewMatrix[i][j];
+      }
+    }
+    skewMatrix = temp;
+  }
+
+  // The affine matrix has a underlying type of coord_t(float) but
+  // we need a double
+  auto reducedDimension = affineMatrix.Ssize() - 1;
+  Mantid::Kernel::DblMatrix affMat(reducedDimension, reducedDimension);
+  for (std::size_t i = 0; i < reducedDimension; i++) {
+    for (std::size_t j = 0; j < reducedDimension; j++) {
+      affMat[i][j] = affineMatrix[i][j];
+    }
```

Comment:
can we refactor this with the `stripMatrix` method or is the type conversion an issue?